File: src/routes/mood.js
const express = require('express');
const multer = require('multer');
const path = require('path');
const { spawn } = require('child_process');

const router = express.Router();
const upload = multer({ storage: multer.memoryStorage(), limits: { fileSize: 6 * 1024 * 1024 } });

const moodMap = {
  happy: 'Happy', joy: 'Happy', surprise: 'Happy', neutral: 'Chill',
  sad: 'Sad', sadness: 'Sad', angry: 'Angry', anger: 'Angry',
  fear: 'Melancholic', disgust: 'Angry', focus: 'Focus',
  no_face: 'Chill'
};

function normalizeMood(label) {
  const key = String(label || '').toLowerCase();
  return moodMap[key] || key.charAt(0).toUpperCase() + key.slice(1) || 'Chill';
}

function fallbackMood() {
  const moods = ['Happy', 'Chill', 'Romance', 'Focus', 'Melancholic'];
  return moods[Math.floor(Math.random() * moods.length)];
}

function callPythonMoodScanner(buffer, mimeType) {
  return new Promise((resolve, reject) => {
    const pythonPath = process.env.PYTHON_PATH || 'python';
    const scriptPath = path.join(__dirname, '..', '..', 'scripts', 'mood_scan.py');
    const child = spawn(pythonPath, [scriptPath], { stdio: ['pipe', 'pipe', 'pipe'] });

    let stdout = '';
    let stderr = '';
    const timer = setTimeout(() => {
      child.kill('SIGTERM');
      reject(new Error('Python mood scanner timed out'));
    }, 15000);

    child.stdout.on('data', chunk => { stdout += chunk.toString(); });
    child.stderr.on('data', chunk => { stderr += chunk.toString(); });
    child.on('error', error => {
      clearTimeout(timer);
      reject(error);
    });
    child.on('close', code => {
      clearTimeout(timer);
      let result = null;
      try {
        result = stdout ? JSON.parse(stdout) : null;
      } catch (error) {
        return reject(new Error(`Invalid Python scanner response: ${stdout || stderr}`));
      }

      if (code !== 0 || result?.error) {
        return reject(new Error(result?.detail || result?.error || stderr || `Python scanner exited with code ${code}`));
      }

      resolve({
        mood: normalizeMood(result.mood),
        confidence: Number(result.confidence || 0),
        rawLabel: result.rawLabel || null,
        provider: result.provider || 'python'
      });
    });

    child.stdin.end(JSON.stringify({
      image: buffer.toString('base64'),
      mimeType: mimeType || 'image/jpeg'
    }));
  });
}

async function callHuggingFace(buffer, mimeType) {
  const token = process.env.HUGGINGFACE_API_KEY;
  const model = process.env.HUGGINGFACE_FACE_MODEL || 'dima806/facial_emotions_image_detection';
  if (!token) return null;

  const response = await fetch(`https://api-inference.huggingface.co/models/${model}`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': mimeType || 'image/jpeg'
    },
    body: buffer
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Emotion model failed: ${response.status} ${text}`);
  }

  const result = await response.json();
  const predictions = Array.isArray(result?.[0]) ? result[0] : result;
  const best = Array.isArray(predictions) ? predictions.sort((a, b) => b.score - a.score)[0] : null;
  if (!best) return null;

  return {
    mood: normalizeMood(best.label),
    confidence: Number(best.score || 0),
    provider: 'huggingface',
    rawLabel: best.label
  };
}

router.post('/analyze', upload.single('face'), async (req, res) => {
  try {
    let buffer = req.file?.buffer;
    let mimeType = req.file?.mimetype;

    if (!buffer && req.body.image) {
      const match = String(req.body.image).match(/^data:(.+);base64,(.+)$/);
      if (match) {
        mimeType = match[1];
        buffer = Buffer.from(match[2], 'base64');
      }
    }

    if (!buffer) return res.status(400).json({ error: 'Upload a face image as multipart field "face" or JSON "image" data URL.' });

    let detected = null;
    try {
      detected = await callPythonMoodScanner(buffer, mimeType);
    } catch (pythonError) {
      try {
        detected = await callHuggingFace(buffer, mimeType);
      } catch (huggingFaceError) {
        detected = null;
      }
      if (!detected) {
        detected = {
          mood: fallbackMood(),
          confidence: 0,
          provider: 'fallback',
          rawLabel: null,
          warning: pythonError.message
        };
      }
    }

    res.json({
      mood: detected.mood,
      confidence: detected.confidence || 0,
      provider: detected.provider || 'fallback',
      rawLabel: detected.rawLabel || null,
      warning: detected.warning || null
    });
  } catch (error) {
    res.status(500).json({ error: 'Mood scan failed', detail: error.message });
  }
});

module.exports = router;

File: backend/mood.py
import base64
import json
import os
import random
import subprocess
from pathlib import Path

import requests
from flask import Blueprint, jsonify, request


mood_bp = Blueprint("mood", __name__, url_prefix="/api/mood")

MOOD_MAP = {
    "happy": "Happy",
    "joy": "Happy",
    "surprise": "Happy",
    "neutral": "Chill",
    "sad": "Sad",
    "sadness": "Sad",
    "angry": "Angry",
    "anger": "Angry",
    "fear": "Melancholic",
    "disgust": "Angry",
    "focus": "Focus",
    "no_face": "Chill",
}


def normalize_mood(label):
    key = str(label or "").lower()
    if key in MOOD_MAP:
        return MOOD_MAP[key]
    return key[:1].upper() + key[1:] if key else "Chill"


def fallback_mood():
    return random.choice(["Happy", "Chill", "Romance", "Focus", "Melancholic"])


def call_python_mood_scanner(image_bytes, mime_type):
    python_path = os.getenv("PYTHON_PATH", "python")
    script_path = Path(__file__).resolve().parents[1] / "scripts" / "mood_scan.py"
    payload = json.dumps({
        "image": base64.b64encode(image_bytes).decode("ascii"),
        "mimeType": mime_type or "image/jpeg",
    })
    result = subprocess.run(
        [python_path, str(script_path)],
        input=payload,
        capture_output=True,
        text=True,
        timeout=15,
        check=False,
    )
    data = json.loads(result.stdout or "{}")
    if result.returncode != 0 or data.get("error"):
        raise RuntimeError(data.get("detail") or data.get("error") or result.stderr or f"Python scanner exited with code {result.returncode}")
    return {
        "mood": normalize_mood(data.get("mood")),
        "confidence": float(data.get("confidence") or 0),
        "rawLabel": data.get("rawLabel"),
        "provider": data.get("provider") or "python",
    }


def call_huggingface(image_bytes, mime_type):
    token = os.getenv("HUGGINGFACE_API_KEY")
    model = os.getenv("HUGGINGFACE_FACE_MODEL", "dima806/facial_emotions_image_detection")
    if not token:
        return None
    response = requests.post(
        f"https://api-inference.huggingface.co/models/{model}",
        headers={"Authorization": f"Bearer {token}", "Content-Type": mime_type or "image/jpeg"},
        data=image_bytes,
        timeout=30,
    )
    response.raise_for_status()
    result = response.json()
    predictions = result[0] if result and isinstance(result[0], list) else result
    best = max(predictions, key=lambda item: item.get("score", 0)) if isinstance(predictions, list) and predictions else None
    if not best:
        return None
    return {
        "mood": normalize_mood(best.get("label")),
        "confidence": float(best.get("score") or 0),
        "provider": "huggingface",
        "rawLabel": best.get("label"),
    }


@mood_bp.post("/analyze")
def analyze():
    try:
        image_bytes = None
        mime_type = None

        uploaded = request.files.get("face")
        if uploaded:
            image_bytes = uploaded.read()
            mime_type = uploaded.mimetype
        else:
            image_data = (request.get_json(silent=True) or {}).get("image")
            if image_data and ";base64," in image_data:
                header, encoded = image_data.split(";base64,", 1)
                mime_type = header.replace("data:", "", 1)
                image_bytes = base64.b64decode(encoded)

        if not image_bytes:
            return jsonify({"error": 'Upload a face image as multipart field "face" or JSON "image" data URL.'}), 400

        warning = None
        try:
            detected = call_python_mood_scanner(image_bytes, mime_type)
        except Exception as python_error:
            warning = str(python_error)
            try:
                detected = call_huggingface(image_bytes, mime_type)
            except Exception:
                detected = None
            if not detected:
                detected = {
                    "mood": fallback_mood(),
                    "confidence": 0,
                    "provider": "fallback",
                    "rawLabel": None,
                }

        return jsonify({
            "mood": detected.get("mood"),
            "confidence": detected.get("confidence") or 0,
            "provider": detected.get("provider") or "fallback",
            "rawLabel": detected.get("rawLabel"),
            "warning": warning if detected.get("provider") == "fallback" else None,
        })
    except Exception as error:
        return jsonify({"error": "Mood scan failed", "detail": str(error)}), 500
